[issue-499] Add document for Pravega catalog (#500)
pravega/flink-connectors: documentation/src/docs/catalog.md

## Basic ideas and building blocks of Pravega Catalog

Pravega itself has terms of *streams* and *scopes* which manage the streaming data, but it does not have the concept of tables and databases. However, these terms are close, for example, if a Pravega stream contains semi-structured data like JSON format, it is feasible to map Pravega streams to Flink tables with the help of a schema registry service. 

[Pravega Schema Registry](https://github.com/pravega/schema-registry) is built for such purpose. It is the registry service built on Pravega that helps store and manage schemas for data stored in Pravega streams, and it also provides a factory of methods to standardize the serialization with built-in support for popular serialization formats in Avro, Protobuf, JSON schemas, as well as custom serialization.

### `PravegaRegistryFormatFactory`

> The serialization specialized in the schema registry involves some extra control bits compared to a standard (de)serialization, so a new table format named `pravega-registry` is developed in the connector to define how to map binary data onto table columns on top of it.

Reviewer: ```suggestion
When using Schema Registry serialization, further information is required in order to describe how to map binary data onto table columns.  A new table format named `pravega-registry` has been added to define this mapping.
```

> **Note:** It should only be used internally in the context of `PravegaCatalog` unless you understand what you are doing. Currently, it only supports Json and avro format without any additional encryption and compression codecs.

Reviewer: ```suggestion
**Note:** The `pravega-registry` format factory should ONLY be used with the `PravegaCatalog`. Currently it supports only Json and Avro formats without any additional encryption and compression codecs.
```

| pravega-registry.uri        | required            | (none)        | String       | Pravega Schema Registry service URI                                                 |
| pravega-registry.namespace  | required            | (none)        | String       | Pravega Schema Registry namespace, should be the same name as Pravega scope         |
| pravega-registry.group-id   | required            | (none)        | String       | Pravega Schema Registry group ID, should be the same name as Pravega stream         |
| pravega-registry.format     | optional            | Avro          | String       | Default format for serialization in table sink, Valid values are 'Json' and 'Avro'  |
| pravega-registry.json.*     | optional            | (none)        | -            | Specification for json format, completely inherited from official Flink Json format factory, refer to this [doc](https://ci.apache.org/projects/flink/flink-docs-stable/docs/connectors/table/formats/json/#format-options) for details                                  |

> Based on Pravega, its schema registry and this table format, a `PravegaCatalog` is built to manage Pravega streams as Flink tables.

Reviewer: ```suggestion
A `PravegaCatalog` is built to manage Pravega streams as Flink tables based on it's schema registry and this table format,
```

> It can map all the streams with its Json/Avro schema registered, and users can directly read from/write to the stream without establishing connection with extra SQL DDL.

Reviewer: ```suggestion
It can map all the streams with its Json/Avro schema registered and users can directly read from/write to the stream without establishing connection with extra SQL DDL.
```

## Pravega as a Catalog

> The `PravegaCatalog` enables users to connect Flink to Pravega streams. The concept mapping between Flink Catalog and Pravega is as following:

Reviewer: ```suggestion
The `PravegaCatalog` enables users to connect Flink to Pravega streams. The following table shows the mapping between Flink Catalog and Pravega terms:
```

| Flink Catalog terms                  | Pravega terms     |
|--------------------------------------|-------------------|
| catalog name (defined in Flink only) | N/A               |
| database name                        | scope name        |
| table name                           | stream name       |

> Currently, `PravegaCatalog` only supports limited `Catalog` methods include:

Reviewer: ```suggestion
Currently `PravegaCatalog` only supports a limited set of `Catalog` methods:
```

```java
// The supported methods by Pravega Catalog
PravegaCatalog.listDatabases();
PravegaCatalog.getDatabase(String databaseName);
PravegaCatalog.databaseExists(String databaseName);
PravegaCatalog.createDatabase(String name, CatalogDatabase database, boolean ignoreIfExists);
PravegaCatalog.dropDatabase(String name, boolean ignoreIfNotExists, boolean cascade);
PravegaCatalog.listTables(String databaseName);
PravegaCatalog.getTable(ObjectPath tablePath);
PravegaCatalog.tableExists(ObjectPath tablePath);
PravegaCatalog.dropTable(ObjectPath tablePath, boolean ignoreIfNotExists);
PravegaCatalog.createTable(ObjectPath tablePath, CatalogBaseTable table, boolean ignoreIfExists);
```

> Only these database and table operations are supported now, views/partitions/functions/statistics operations are NOT supported in `PravegaCatalog`.

Reviewer: ```suggestion
Only these database and table operations are currently supported. Views/partitions/functions/statistics operations are NOT supported in `PravegaCatalog`.
```

execution:
  ...
  current-catalog: pravega_catalog  # set the PravegaCatalog as the current catalog of the session
  current-database: scope1
 
catalogs:
  - name: pravega_catalog
    type: pravega
    controller-uri: tcp://localhost:9090
    schema-registry-uri: http://localhost:9092
    default-database: scope1
```

After that, you can operate Pravega scopes and streams with SQL commands. Here are some examples.
```sql
-- List all the scopes
SHOW DATABASES;

-- Create a scope
CREATE DATABASE scope2 WITH (...);

-- Delete a scope, if 'CASCADE' is followed, all the streams will also be dropped.
DROP DATABASE scope2;

-- List all the streams with schema registered in the scope
SHOW TABLES;

-- Create a stream and register the schema
CREATE TABLE mytable (name STRING, age INT) WITH (...);

-- Delete a stream, the schema group will also be deleted 
DROP DATABASE scope2;

-- Scan/Query the stream `test_table` as a table
SELECT * FROM pravega_catalog.scope1.test_table;
SELECT count(DISTINCT name) FROM test_table;

-- Write rows/Copy a table into the stream `test_table`
INSERT INTO test_table VALUES('Tom', 30);
INSERT INTO test_table SELECT * FROM mytable;
```

## Useful Flink links

> Users can check [Flink Table catalogs docs](https://ci.apache.org/projects/flink/flink-docs-stable/dev/table/catalogs.html) to learn more about the general idea of catalog and detailed operations.

Reviewer: ```suggestion
See [Flink Table catalogs docs](https://ci.apache.org/projects/flink/flink-docs-stable/dev/table/catalogs.html) for more information on the general Catalog concepts and more detailed operations.
```
